src/mgmt/cli_mgmt.c

```diff
@@ -1271,6 +1272,158 @@ dc_mgmt_tm_register(const char *sys, const char *jobid, key_t shm_key, uid_t *ow
 	return rc;
 }
 
+int
+dc_mgmt_pool_list(tse_task_t *task)
+{
+	daos_mgmt_pool_list_t     *args;
+	d_rank_list_t             *ms_ranks;
+	struct rsvc_client         ms_client;
+	crt_endpoint_t             ep;
+	crt_rpc_t                 *rpc = NULL;
+	crt_opcode_t               opc;
+	struct mgmt_pool_list_in  *in  = NULL;
+	struct mgmt_pool_list_out *out = NULL;
+	struct dc_mgmt_sys        *sys;
+	int                        rc, pidx;
+
+	args = dc_task_get_args(task);
+	if (args->npools == NULL) {
+		D_ERROR("npools argument must not be NULL");
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	rc = dc_mgmt_sys_attach(args->grp, &sys);
+	if (rc != 0) {
+		DL_ERROR(rc, "cannot attach to DAOS system: %s", args->grp);
+		D_GOTO(out, rc);
+	}
+
+	ms_ranks = sys->sy_info.ms_ranks;
+	D_ASSERT(ms_ranks->rl_nr > 0);
+
+	rc = rsvc_client_init(&ms_client, ms_ranks);
+	if (rc != 0) {
+		DL_ERROR(rc, "failed to init ms client");
+		D_GOTO(out_grp, rc);
+	}
+
+	ep.ep_grp = sys->sy_group;
+	ep.ep_tag = daos_rpc_tag(DAOS_REQ_MGMT, 0);
+	opc       = DAOS_RPC_OPCODE(MGMT_POOL_LIST, DAOS_MGMT_MODULE, DAOS_MGMT_VERSION);
+
+rechoose:
+	rc = rsvc_client_choose(&ms_client, &ep);
+	if (rc != 0) {
+		DL_ERROR(rc, "failed to choose MS rank");
+		D_GOTO(out_client, rc);
+	}
+
+	rc = crt_req_create(daos_task2ctx(task), &ep, opc, &rpc);
+	if (rc != 0) {
+		DL_ERROR(rc, "crt_req_create(MGMT_POOL_LIST) failed");
+		D_GOTO(out_client, rc);
+	}
+
+	in          = crt_req_get(rpc);
+	in->pli_grp = (d_string_t)args->grp;
+	/* If provided pools is NULL, caller needs the number of pools
+	 * to be returned in npools. Set npools=0 in the request in this case
+	 * (caller value may be uninitialized).
+	 */
+	if (args->pools == NULL)
+		in->pli_npools = 0;
+	else
+		in->pli_npools = *args->npools;
+
+	/* Now fill in the client credential for the pool access checks. */
+	rc = dc_sec_request_creds(&in->pli_cred);
+	if (rc != 0) {
+		DL_ERROR(rc, "failed to obtain security credential");
+		D_GOTO(out_put_req, rc);
+	}
+
+	D_DEBUG(DB_MGMT, "req_npools=" DF_U64 " (pools=%p, *npools=" DF_U64 "\n", in->pli_npools,
+		args->pools, *args->npools);
+
+	crt_req_addref(rpc);
+	rc = daos_rpc_send_wait(rpc);
+	if (rc != 0) {
+		DL_ERROR(rc, "rpc send failed");
+		crt_req_decref(rpc);
+		goto rechoose;
```

Comment:
I implemented this client API to follow the pattern used by other rsvc clients, where the request is retried forever, or at least until the user decides to cancel it. As I recall from a conversation with @liw about this pattern, the idea is that in a large-scale distributed system, it may be difficult to distinguish from the client's perspective from a truly broken system vs. a system that is in the process of initializing or is otherwise not ready yet.

As we have seen on the `dmg` side of things, it's challenging to choose a timeout value that works reliably in all cases.

Reply:
Sounds logical.  Thanks for the explanation.

Reply:
@knard-intel, for the engine side, we have introduced a deadline via `dsc_pool_svc_call`; for the client side, since the DAOS API does not have the concept of timeout and deadline, it's okay to continue using infinite retries, I think.

Reply:
@liw thanks for the detailed explanation

Reply:
@knard-intel, np, it was a good question.